### DISCUSSION TITLE
VCR Proxy Server

### DISCUSSION THREAD

**Discussion by myronmarston** [vcr/vcr#863] [Ideas]
I've had the idea for a while to build a VCR proxy server, either as an extra gem that builds on top of VCR (e.g. `vcr-proxy`) or as an extra piece within the VCR gem itself.  There are a lot of different directions this could go, so I want to solicit feedback from VCR users before beginning any work on it.

For now, if you have any feedback on this, please add your thoughts at the bottom of this document.
## Problems the proxy server may solve
- As @jonleighton [has pointed out](https://twitter.com/jonleighton/status/227340874395295745), VCR relies pretty extensively on monkey patching to do its thing, at least if you use either FakeWeb or WebMock (as most people do). VCR can operate virtually monkey-patch free if you use it with Faraday, Excon or Typhoeus, as those libraries provide public APIs that VCR can use.  It'd be great to remove the dependency on monkey-patched code.
- VCR only works for ruby, and it only works for some HTTP libraries (although, at this point, it works for most ruby HTTP libraries). There have been [frequent](http://stackoverflow.com/questions/7802499/node-js-testing-with-vcr) [questions](http://stackoverflow.com/questions/7781881/java-alternative-for-rubys-vcr) asking about ports of VCR in other languages since it's such a useful tool for testing HTTP-dependent code.

I think a VCR proxy server is a promising potential way to solve both of these problems.  The proxy server could be used by any HTTP client, including those in other languages, and it wouldn't need any monkey patching at all.

Currently, VCR has some tooling that makes it possible to build your own proxy server (e.g. the VCR rack middleware), but that still takes a fair amount of work.  I'd like to make it trivial to use a VCR proxy server.
## Open Questions / Discussion
1. How should the proxy server know what cassette to use?  There are two basic ways we could do it here:
   1. **Internal** configuration -- when the proxy server receives a request it can select a cassette on the fly, based on some criteria or a callback or whatever.  This is similar to how you can use the [around_http_request hook](https://www.relishapp.com/myronmarston/vcr/v/2-2-4/docs/hooks/around-http-request-hook) to use a cassette for a particular HTTP request based on the HTTP request using a block callback.
   2. **External** configuration -- this is the typical way VCR is used: in your test suite, external from the HTTP request itself, you insert a cassette, and any HTTP requests that are made while the cassette is inserted use it.
2. What should the API to start the proxy server be?  Again, there's a couple possibilities here:
   1. **An executable** -- The gem could ship with an executable that starts the the proxy server. It would presumably accept options for things like what port to run on, as well as the path to a configuration file (which would probably be a simple be a ruby file).
   2. **A ruby class** -- Something like `VCR::ProxyServer.start(:port => 5000)`.
3. Should this be part of the VCR gem or a separate gem?
4. What HTTP server libraries should the proxy server use?  Just Rack/WeBrick? Thin? Something else?

I'm not sure which way the proxy server could work.  It seems to me that in the test suite of a ruby project, a developer would probably want to launch the proxy server in a separate thread, and control the cassettes it uses externally (similar to how VCR is used now). In the context of a non-ruby project, it would probably be more convenient to launch the proxy server in its own process using an executable, and pass the executable the path to a ruby file that configures cassettes based on a callback that receives the HTTP request as an argument.

Is one way more desirable than the other?  Or is there a way to meet both needs?  Are there other use cases I'm not thinking of here?

Please respond below with your thoughts!

**Comment by myronmarston** [comment]
From @jrochkind:
- Hmm. This sounds cool, but also sounds potentially complicated. One of the nice things about VCR is how danged simple it's implementation is, for what it does. If something is going wrong, it's relatively easy to debug, even if it is a bug in VCR or in VCR's integration with the HTTP libraries. The one unpleasant part, as you note, is the monkey patching, but even still. The proxy idea could do some really cool stuff (esp for other languages), but at the cost of a lot more moving parts.
  - In the context of a ruby app, where VCR's monkey patching works properly (both big assumptions), the proxy method seems like it would take a lot of work to be as _easy_ to set up and use and current VCR.  Having to manually start a seperate proxy process and/or do VCR configuration and casette control somewhere _other_ than your inline testing code -- would def not be as good an experience as present. Part of why people want "something like VCR for other languages" is because of how darn easy to use and rock solid VCR is -- you just learn a couple API calls, add em to your test code, and everything JUST WORKS, it's awesome. So I would def vote for 'auto launch in a seperate thread without me even knowing it happened, plus cassette control similar to now inline with tests' continuing to be a possibility at least in ruby. But really I'm not sure I'd use the proxy method at all unless I had to (the old way of monkeypatch and/or hook into HTTP clients no longer being supported, which would make me sad). 
  -  So I guess I'd vote for it being a separate gem, with VCR still being supported in it's present manner of working too (which certainly could be done with the proxy being an option in the present gem too, but a seperate gem creates that psychological and code base seperation to make it more likely to stay entirely de-coupled).

**Comment by myronmarston** [comment]
From @mislav:

It would definitely be interesting to have a proxy server because of reasons outlined above. However here are some drawbacks I can think of:
- starting/stopping the server: a separate thread in Ruby apps would not work for all HTTP adapters, namely Patron. So you'd have to use process management. But process management is a bit slower (wait times for web server to start up/shut down) and might not be perfectly portable to Windows (just guessing after not developing on Windows for years). Our test suites would require additional wrapper scripts to handle this. I'm in favor of wrapper scripts in general, but they do add complexity to a workflow that was more straightforward in the way we use VCR now.
- dynamic configuration/hooks: with VCR in Ruby we have access to all its configuration and can create hooks. If VCR was running in a separate process, we couldn't dynamically alter its configuration without some inter-process communication. Also, how would non-Ruby users write the hooks, which are for me the killer feature of VCR?

**Comment by FDM** [comment]
I was about to start investigating doing something like this too.  I use selenium or webkit headless for browser testing as my application is javascript heavy and uses many different 3rd party services such as Twitter, Facebook, Google +1 and more.
I would like to be able to record a test, then use this recording for future tests, irrespective of what makes the http(s) request - browser, webkit headless or ruby.
The only way I can think of to do this is using a proxy.
I have used VCR and having a proxy with VCR built in sounds absolutely perfect to me.
I would guess there would need to be a simple API between the test suite and the proxy to do things like changing cassettes.  This could be kept really simple so not to over complicate things.
Whilst I dont speak for everyone when I say this, performance is not a key concern to me as the test suite is already slow due to the nature of browser testing.

**Comment by sdhull** [comment]
I guess I still don't quite see how this solves the problem in Selenium or Webkit-Capybara.  Is it somehow possible to tell Selenium or Webkit-Capybara to proxy the loading of assets through another server?

Generally, if I have a tag in my page like:

```
<script src="http://someotherserver.com/foo.js" />
<img src="http://someotherserver.com/logo.png" />
```

How can this be solved by a vcr proxy server?

FWIW, I'd be very interested in a solution to this problem.  As we added webkit-capybara to our test suite to test js interactions and such, we realized that we added back in dependency on external services to our test suite, which is bad news indeed.  The day Mixpanel went down, our suite was red all day (artificially, I would say).

Furthermore, having personally re-written the `webmock` support for `em-http-request` when the internals completely changed and all the old monkey-patching no longer worked, I would very much be interested in some way to avoid these monkey patches and instead define clear APIs and hooks that VCR (and webmock, for that matter) could leverage to do our business and be immune to internal changes in these http libraries.

**Comment by myronmarston** [comment]
@sdhull -- good questions.  I don't have a great answer for you.  A VCR proxy server may help with some problems but won't solve all problems.  I'm not sure if it'd help in your case or not.

**Comment by opsb** [comment]
I really like this idea. For configuring the cassette/record options you could add some query params onto the address for the proxy e.g. at the start of your test you set the proxy on capybara-webkit to http://vcrproxy.dev?cassette=create_account&record_mode=once. See https://github.com/thoughtbot/capybara-webkit/pull/172 for info on capybara-webkits http proxy option.

**Comment by mislav** [comment]
@opsb: Can proxy servers receive path and query params? I thought proxy configuration was only host+port+scheme (and optionally Basic auth).

**Comment by opsb** [comment]
I have to admit I wasn't sure about that. An alternative though would be to use subdomains, this would limit you to only specifying the cassette but I find that I set the options globally anyway.

**Comment by mislav** [comment]
A much better option for configuring the proxy would be command-line arguments when you start the process.

**Comment by oesmith** [comment]
I've just released a library for rspec+capybara that proxies and stubs HTTP requests.  It might be worth a look (if only to see the approach I've taken to the implementation)

https://github.com/oesmith/puffing-billy

**Comment by jweissman** [comment]
Just here because I'd love to see this! Would be happy to help out, too, if there are tasks around this that are commensurate with my skill set.

**Comment by 23tux** [comment]
This would be very cool! I just did some research with creating a proxy and recording that with VCR (for a FeedZirra, because they use curl multimode). This worked fine, but then I tried to run Poltergeist through that VCR proxy, and it doesn't work.

So, if you guys need some help, I would be happy to contribute!

EDIT:
I played around a little bit, 'cause I would need VCR for recording ajax requests from phantomJS. Maybe some of you can help: http://stackoverflow.com/questions/13039251/record-phantomjs-ajax-calls-with-vcr-inside-capybara

**Comment by croaky** [comment]
There may be some ideas in here:

https://gist.github.com/079a31bc4f46f50bf485

**Comment by 23tux** [comment]
@croaky aren't you just stubbing out facebook calls with fake classes? Thats also an option, but I think it would be more comfortable if you just could record the facebook calls from the real api and playback them (and as I said above, even from javascript through a proxy)

**Comment by 23tux** [comment]
Is anybody currently working on this? I haven't found a solution yet

**Comment by myronmarston** [comment]
@23tux -- as far as I know, no one has taken a stab at this yet.  I haven't had the time and probably won't anytime soon.  I'd still like to see this happen, though.

**Comment by bowsersenior** [comment]
@myronmarston This feature would be very useful, in my opinion. Have you seen https://github.com/pmoran/nox-vcr ? Seems very similar.

**Comment by bowsersenior** [comment]
Also worth a look:
- https://github.com/hooroo/nox
- https://github.com/hooroo/nox-ruby

**Comment by 23tux** [comment]
@bowsersenior cool, nice way to do such a proxy ;) But if you're doing so, you have to have access to all calls, so that you can rewrite the url to /request. I think thats hard if you are dealing with external javascript libraries, that are doing ajax calls. Here it would be better to just proxy them and record the response.

I'm now a step further: I'm able to record all HTTP requests with VCR, as long as the application can be configured with a proxy option. The only drawback is HTTPS, because the calls gets proxied through with that `CONNECT` flag (whatever). If we want to record HTTPS traffic, we could generate a certificate on the fly (of course not valid, but webkit or phantomjs can be configured to ignore ssl errors in testing mode), access the content, let VCR record it and so on.

Unfortunately I have no idea how to do this. It should be some kind of man in the middle proxy, but I haven't found one using ruby. If some of have more information, please let me know!

**Comment by taqtiqa-mark** [comment]
@23tux, do you have a repo we can look at.  It looks like this type of VCR recording proxy might allow me to use VCR with FakeFS. Right now that is proving a tricky combination. 

People might be interested in Ilya Grigorik's [em-proxy](https://github.com/igrigorik/em-proxy) "EventMachine Proxy DSL for writing high-performance transparent / intercepting proxies in Ruby"

@myronmarston, any thoughts on em-proxy, i.e. where to drop in the VCR part and what the VCR might look like. If I read @23tux 's SO code correctly, everything gets recorded under one cassette name - correct?

**Comment by 23tux** [comment]
I had a look at the em-proxy, this might be a good solution, but it doesn't support HTTPS / man in the middle. I will dive deeper in this topic next weekend, I'll then put a working example of the http version on a repo and post it here.

**Comment by taqtiqa-mark** [comment]
@23tux, I think the path of least pain for a https proxy will be [em-proxy](https://github.com/igrigorik/em-proxy)+[eventmachine-le](https://github.com/ibc/EventMachine-LE)

Whether em-le is a drop in replacement from the point of view of em-proxy is something you'll only discover by trying it out.
I do know that em-le has pulled in patches related to ssl support, so you'll likely save a ton of time by trying em-proxy on top of em-le.

Look forward to seeing anything you mange to get working.

Hope that helps?

**Comment by brutuscat** [comment]
Here is a nice example of using stunnel to place a proxy as MITM http://www.fishnetsecurity.com/6labs/blog/ssl-relay-proxy-creative-solution-complex-issue 

I though it may help ;)

**Comment by oesmith** [comment]
FWIW, [puffing-billy](https://github.com/oesmith/puffing-billy) v0.2.0 now has the ability to persist its cache and operate like a pretty rudimentary version of VCR.

**Comment by myronmarston** [comment]
@oesmith -- that's great.  If you decide you want a fuller set of VCR features, w/o needing to write them yourself, I think it would be pretty trivial to use VCR in puffing billy to do the recording and replaying.

**Comment by 23tux** [comment]
Hey everyone, I took some time and tried to make a working (but very basic) example of a VCR proxy (that also handles HTTPS when you disable the warnings in your client). I would be very happy when someone contributes or gives feedback, and I already made some issues that should be solved. Here is the repo https://github.com/23tux/vcr_proxy

**Comment by oesmith** [comment]
@myronmarston reading through one of puffing-billy's pull requests today, I realised that using puffing-billy with VCR in the same project will actually route all un-stubbed requests to VCR. I'm using EM::HttpRequest for proxy requests, and I guess VCR hooks into that!

**Comment by dkubb** [comment]
Here's one example of a vcr proxy I wrote once for testing Feedzirra: https://gist.github.com/dkubb/1434673

**Comment by mislav** [comment]
@23tux 

> I took some time and tried to make a working (but very basic) example of a VCR proxy (that also handles HTTPS when you disable the warnings in your client). I would be very happy when someone contributes or gives feedback, and I already made some issues that should be solved. Here is the repo https://github.com/23tux/vcr_proxy

Heh, [great minds think alike](https://github.com/mislav/dotfiles/blob/167bb068f4e05e2cbcbd5a68318401725f31db21/bin/proxy) (although my proxy code is not related to VCR; it's just for logging HTTP requests). I've also spawned an additional WEBrick server in the background to act as a MITM for SSL traffic. From your code, however, it seems that you're spawning a new SSL server on every CONNECT. Is that necessary?

**Comment by elgalu** [comment]
+1 this feature would be awesome!

**Comment by burgalon** [comment]
I've written a small sinatra_proxy which helps VCR too hook into a rack up.
https://gist.github.com/burgalon/e1e95c26f6e1bd762909

**Comment by mcfiredrill** [comment]
I think it would be great if VCR had a proxy like puffing-billy, or even if they could integrate together easier.
Getting VCR and puffing billy to work together is not super simple, and puffing billy's caching doesn't have all the features that VCR cassettes do.

**Comment by brutuscat** [comment]
Have you guys heard of [mountebank](http://www.mbtest.org/)

> mountebank is the first open source tool to provide cross-platform, multi-protocol test doubles over the wire. Simply point your application under test to mountebank instead of the real dependency, and test like you would with traditional stubs and mocks.

It seems that should be fairly "simple" to wire VCR here.

**Comment by krainboltgreene** [comment]
This is more relevant now that serverless exists.

**Comment by alexgenco** [comment]
This issue seems pretty stale, but FWIW we would very much like this feature, and the landscape for alternatives is surprisingly bare (we're using [wiremock standalone](http://wiremock.org/docs/running-standalone/), which is possibly deprecated, and is missing support for something like `record: :all` without weird hacks).

Are there any other vetted alternatives out there?

**Comment by dmitry** [comment]
@alexgenco what about puffing-billy, it's now mature. If you can setup proxy on your client, then it will work.

**Comment by ryanong** [comment]
puffing billy is a beast that already includes it's own request stubbing. It would probably be best to use `async-http` as a proxy and then use vcr to capture requests

https://github.com/socketry/async-http/blob/master/spec/async/http/proxy_spec.rb is a good example of it working
